lectures/11.rendering_2.md

```diff
@@ -57,259 +57,968 @@ We are recording, aren't we?
-and finally blended into the frame buffer.
+Definitions
+* Latency
+ *  Time to finish a fixed task
+  *   33 or 16 ms frame ( 30 or 60 fps)
+  *   15 ns for inserting an element into a queue
+Throughput
+* Tasks per unit time
+  *  16 GB/s memory transfer
+  *  4 Megapixels/s shading samples
 
 --- VERTICAL SLIDE ---
+Processor history
 
-![Pipeline](resources/10.rendering/pipeline.jpg)
+* Moore's law
+ * The number of transistors can be doubled every (1.5 years) (1969)
+* Performance
+  * Transistor performance. 
+   * Improves linearly as size scales down
+  * Wires scale poorly. 
+   * Delays are increased, signals are lost. Relays are installed
+   * 2 km of wiring on AMD Rysen
 
---- NEXT SLIDE ---
+--- VERTICAL SLIDE ---
+Processor history
 
-## GPU Architecture
+* TDP – Thermal design power
+  * Determines cooling requirements. Measured in watts (ex. 15W, 100W, 250W)
+  * Power supply and cooling match and exceed TDP
+  * Power = ½ Capacitive Load * Voltage ^ 2 * Frequency
+  * Capacitive Load – function of number of transistors
+  * Voltage – depends on transistors
+  * Frequency can be adjusted
 
-![GPUArch](http://www.keremcaliskan.com/wp-content/uploads/2011/01/CPU-GPU-Structures1.png)
+--- VERTICAL SLIDE ---
+Obstacles to increasing frequency
+
+![Obstacles](http://www.gotw.ca/images/CPU.png)
 
 --- VERTICAL SLIDE ---
+Memory barrier - DRAM
 
-## throughput vs latency
+| Year | DRAM Growth | Capacity |
+| ---- | ---- | ---- |
+|1990 | 60%/year | 4x/3 years |
+|1996 | 60%/year | 4x/3 years |
+|2003 | 40-60%/year | 4x/3-4 years |
+|2007 | 40%/year | 2x/2 years |
+|2011 - | 25-40%/year | 2x/2-3 years |
 
---- NEXT SLIDE ---
+--- VERTICAL SLIDE ---
+Capital for research
 
-## GPU pipeline
+![renderdoc](http://www.yole.fr/iso_album/illus_annualcapitalexpenditure_standalonedram_yole_june2019_(430x278).jpg)
 
 --- VERTICAL SLIDE ---
+Memory barrier - speed
 
-![GPU Pipeline](https://traxnet.files.wordpress.com/2011/07/shader_full.jpeg)
+![Pipeline](resources/10.rendering/memory_barrier_speed.png)
 
---- NEXT SLIDE ---
+* 1980: RAM latency ~ 1 cycle
+* 2011: RAM latency ~ 150 - 400 + cycles
 
-## The hardware device
+Sin(x) values in a table are faster in 1980
 
-Abstracts the GPU for you
+--- VERTICAL SLIDE ---
+* CPU
+  * Low latency, low throughput ( ns, < 100 GB/s )
+* GPU
+  * High latency, high throughput (ms, > 100 GB /s )
+
+| | AMD Athlon II x4 | Radeon 5750 |
+| ---- | ---- | ---- |
+|TDP|95W| 86W|
+|FLOPS| 30 GFLOPS| 1008 GFLOPS|
+|Transistors| 300 M (768PH)| 1040 M|
+|Bandwith GB/s| 9 ( 4 OpenCL )| 70 ( 40 OpenCL )|
 
-* DirectX
-* OpenGL
-* Console-specific
-* Console on nonconsoles - DX12, Vulcan, Metal
+--- VERTICAL SLIDE ---
+CPU Components
+* ALU
+  * compute
+* Fetch
+  * instructions stream, microops
+* Execution context
+  * registers, where is the program
+* Branch predictor
+  * conditions, if, else, while, for
 
 --- VERTICAL SLIDE ---
+CPU Components
 
-## The shader model
+* Out of order logic

```

Comment:
Consider animating some of these bullets - I've noticed some students tend to start reading the full text on the slide as it's shown and get distracted instead of listening